src/test/java/com/teragrep/rlp_03/client/ClientTest.java

```diff
@@ -0,0 +1,198 @@
+ *
+ * Licensee must indemnify licensors and authors for any liability that these
+ * contractual assumptions impose on licensors and authors.
+ *
+ * To the extent this program is licensed as part of the Commercial versions of
+ * Teragrep, the applicable Commercial License may apply to this file if you as
+ * a licensee so wish it.
+ */
+package com.teragrep.rlp_03.client;
+
+import com.teragrep.rlp_03.*;
+import com.teragrep.rlp_03.context.channel.PlainFactory;
+import com.teragrep.rlp_03.context.channel.SocketFactory;
+import com.teragrep.rlp_03.delegate.DefaultFrameDelegate;
+import org.junit.jupiter.api.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.charset.StandardCharsets;
+import java.util.AbstractMap;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+public class ClientTest {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClientTest.class);
+    private Server server;
+    private Thread serverThread;
+    private final int port = 23601;
+    private ExecutorService executorService;
+
+    @BeforeAll
+    public void init() {
+        executorService = Executors.newSingleThreadExecutor();
+        ServerFactory serverFactory = new ServerFactory(
+                executorService,
+                new PlainFactory(),
+                () -> new DefaultFrameDelegate((frame) -> LOGGER.debug("server got <[{}]>", frame.relpFrame()))
+        );
+        Assertions.assertAll(() -> {
+            server = serverFactory.create(port);
+
+            serverThread = new Thread(server);
+            serverThread.start();
+
+            server.startup.waitForCompletion();
+        });
+    }
+
+    @AfterAll
+    public void cleanup() {
+        server.stop();
+        executorService.shutdown();
+        Assertions.assertAll(() -> serverThread.join());
+    }
+
+    private class RunnableEventLoop implements Runnable, Closeable {
+
+        private final AtomicBoolean stop;
+        private final EventLoop eventLoop;
+
+        RunnableEventLoop(EventLoop eventLoop) {
+            this.stop = new AtomicBoolean();
+            this.eventLoop = eventLoop;
+        }
+
+        @Override
+        public void close() throws IOException {
+            stop.set(true);
+            eventLoop.wakeup();
+        }
+
+        @Override
+        public void run() {
+            LOGGER.debug("running eventLoop <{}>", eventLoop);
+            try {
+                while (!stop.get()) {
+                    eventLoop.poll();
+                }
+            }
+            catch (IOException e) {
+                throw new RuntimeException(e);
+            }
```

Comment:
Maybe Assertions.assertDoesntThrow this, or at least Assertions.fail("IOEXception happened") in the catch if it can't be realistically modified

Reply:
will be fixed later #120